### PR TITLE
Delete Labels & IssueLabels on Repo Delete too

### LINKED ISSUE #15036: [BUG] Delete repo do not delete related issue_label entrys
tested with gitea 1.14.0+dev-877-g0e5c6c449

reproduce:
1. create repo with 1 issue and 1 label
1. assign label to issue
1. delete repo
1. exec: `select COUNT(issue_label.id) from issue_label LEFT JOIN issue ON issue_label.issue_id = issue.id LEFT JOIN label ON issue_label.label_id = label.id WHERE issue.id IS NULL OR label.id IS NULL;`

### PR DESCRIPTION
**Pull request by 6543**
close #15036

### CODE REVIEW

File: models/issue_label.go [hunk 1]
```diff
@@ -780,3 +780,15 @@ func DeleteIssueLabel(issue *Issue, label *Label, doer *User) (err error) {
 
 	return sess.Commit()
 }
+
+func deleteIssueLabelsByRepoID(sess Engine, repoID int64) error {
```
Comment (6543) [inline models/issue_label.go]:
```suggestion
func deleteLabelsByRepoID(sess Engine, repoID int64) error {
```

File: models/repo.go [hunk 2]
```diff
@@ -1475,28 +1475,32 @@ func DeleteRepository(doer *User, uid, repoID int64) error {
 		return fmt.Errorf("deleteBeans: %v", err)
 	}
 
+	if err := deleteIssueLabelsByRepoID(sess, repoID); err != nil {
```
Comment (6543) [inline models/repo.go]:
```suggestion
	if err := deleteLabelsByRepoID(sess, repoID); err != nil {
```

File: models/consistency.go [hunk 3]
```diff
@@ -224,6 +224,24 @@ func DeleteOrphanedLabels() error {
...
+// CountOrphanedIssueLabels return count of IssueLabels witch have no label behind anymore
+func CountOrphanedIssueLabels() (int64, error) {
+	return x.Table("issue_label").
+		Join("LEFT", "label", "issue_label.label_id = label.id").
+		Where(builder.IsNull{"label.id"}).Count()
+}
+
+// DeleteOrphanedIssueLabels delete IssueLabels witch have no label behind anymore
+func DeleteOrphanedIssueLabels() error {
+
```
Comment (6543) [inline models/consistency.go]:
```suggestion
```

File: models/repo.go [hunk 4]
```diff
@@ -1475,28 +1475,32 @@ func DeleteRepository(doer *User, uid, repoID int64) error {
 		return fmt.Errorf("deleteBeans: %v", err)
 	}
 
+	if err := deleteLabelsByRepoID(sess, repoID); err != nil {
```
Comment (6543) [inline models/repo.go:1479]:
```suggestion
	// Delete Labels and related objects
	if err := deleteLabelsByRepoID(sess, repoID); err != nil {
```

File: models/migrations/v176.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if err := sess.Sync2(new(IssueLabel)); err != nil {
+		return fmt.Errorf("Sync2: %v", err)
+	}
+
+	if _, err := sess.In("id", builder.Select("issue_label.id").From("issue_label").
+		Join("LEFT", "label", "issue_label.label_id = label.id").
```
Comment (zeripath) [inline models/migrations/v176.go]: Mysql is not going to like this:

            	            	migrate: do migrate: Error 1093: You can't specify target table 'issue_label' for update in FROM clause

The reference to issue_label needs to have an alias. 

This is exactly the same problem I faced in my PRs around issue_label

Comment (6543) [reply models/migrations/v176.go]: the stupid thing is: "it works on my pc" ... just fine :/

**Comment by 6543** [comment]
@zeripath this is just the SQL query converted from builder into raw SQL ... + an `AS` ... witch worked on my PC too

**Comment by lafriks** [comment]
MySQL migrations fail with:
```
Unknown column 'conflicted_files' in 'field list'
```

File: models/migrations/v176.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+	if err := sess.Sync2(new(IssueLabel)); err != nil {
+		return fmt.Errorf("Sync2: %v", err)
+	}
+
+	if _, err := sess.Exec(`DELETE FROM issue_label WHERE issue_label.id IN (
+		SELECT il.id FROM issue_label AS il
+			LEFT JOIN label ON il.label_id = label.id
+		WHERE
+			label.id IS NULL)`); err != nil {
```
Comment (6543) [inline models/migrations/v176.go]:
```suggestion
			label.id IS NULL) AS ill`); err != nil {
```

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by lafriks** [comment]
conflicts ;)

**Review by zeripath — APPROVED** [review]
Probably we need Xorm to be able to set aliases in this way with builder

**Review by lafriks — APPROVED** [review]

**Pull request merged**